### PR TITLE
Iterative improvements to scenario management and creation

### PR DESCRIPTION
**Pull request by brysonjbest**
- [x] Add ability to edit scenarios without drastically modifying the existing ui.
- [x] Edits take place in the simulate inputs tab.
- [x] In order to handle null values correctly, add a "clear input" button to inputs that properly resets a value to a null value. This helps when re-writing a rule if a variable is no longer in use or has been renamed, the user no longer needs to completely re-create scenarios, but simply edit the fields.
- [x] Add a custom datepicker for inputs based on property name.
- [x] Refactor currency formatting.
- [x] Ability to sort all columns in scenario testing.
- [x] Ability to filter all columns in scenario testing.
- [x] Reformat scenario testing table for dynamic scrolling.

### CODE REVIEW

File: app/components/ScenariosManager/ScenariosManager.tsx [hunk 1]
```diff
@@ -40,13 +40,18 @@ export default function ScenariosManager({
   resultsOfSimulation,
 }: ScenariosManagerProps) {
   const [resetTrigger, setResetTrigger] = useState<boolean>(false);
+  const [activeKey, setActiveKey] = useState<string>(showAllScenarioTabs ? "3" : "1");
```
Comment (timwekkenbc) [inline app/components/ScenariosManager/ScenariosManager.tsx]: Rather than using string values like "3" and "1", you should use an enum so that it's clear what tab you're referring to. 
Example:
```
enum tabs {
    ScenariosTab,
    InputsTab,
    ResultsRab,
    CSVTab
}
```

File: app/components/ScenariosManager/ScenariosManager.tsx [hunk 2]
```diff
@@ -57,20 +62,24 @@ export default function ScenariosManager({
...
       <ScenarioViewer
-        scenarios={scenarios}
+        scenarios={activeScenarios}
+        jsonFile={jsonFile}
         setSimulationContext={setSimulationContext}
         resultsOfSimulation={resultsOfSimulation}
         runSimulation={runSimulation}
         rulemap={rulemap}
         editing={isEditing}
+        setActiveKey={setActiveKey}
```
Comment (timwekkenbc) [inline app/components/ScenariosManager/ScenariosManager.tsx]: Super minor nitpick, but I find "setActiveKey" to be a little bit non-descriptive. Like if I'm looking at a subcomponent and I see `setActiveKey?.("2");` I don't really understand that it's changing the tab of the parent component

File: app/components/ScenariosManager/ScenarioViewer/ScenarioViewer.tsx [hunk 3]
```diff
@@ -59,6 +67,22 @@ export default function ScenarioViewer({
...
   };
 
+  const handleEditScenario = (scenario: Scenario) => {
+    setManageScenarios(false);
+
+    const scenarioInputs = scenario.variables.reduce((acc, variable) => {
+      acc[variable.name] = variable.value;
+      return acc;
+    }, {} as Record<string, any>);
+    const editScenarioInputs = { ...scenarioInputs, ...{ rulemap: true } };
```
Comment (timwekkenbc) [inline app/components/ScenariosManager/ScenarioViewer/ScenarioViewer.tsx]: `{ ...scenarioInputs, rulemap: true }` is the cleaner version of this. I've done this same thing before though haha

File: app/components/ScenariosManager/ScenarioGenerator/ScenarioGenerator.tsx [hunk 4]
```diff
@@ -48,31 +63,45 @@ export default function ScenarioGenerator({
...
+        await updateScenario(newScenario, existingScenario._id);
+      } else {
+        await createScenario(newScenario);
+      }
+      setScenarioName?.("");
+      setSimulationRun(false);
+      setEditingScenario(false);
+      setSimulationContext(newScenario);
+      updateScenarios();
+      setActiveKey?.("1");
```
Comment (timwekkenbc) [inline app/components/ScenariosManager/ScenarioGenerator/ScenarioGenerator.tsx]: All of the logic here for preparing a tab that isn't this one to be switched to (I think that's what happening here) should ideally not live in this subcomponent, but either in the other component or the shared parent component.

File: app/components/ScenariosManager/ScenarioGenerator/ScenarioGenerator.tsx [hunk 5]
```diff
@@ -48,31 +63,45 @@ export default function ScenarioGenerator({
...
+    if (scenarios.find((scenario) => scenario.title === scenarioName)) {
+      setEditingScenario(true);
+    }
     runSimulation();
     setSimulationRun(true);
   };
 
   useEffect(() => {
     setSimulationRun(false);
+    const editScenario = { ...simulationContext, ...{ rulemap: true } };
```
Comment (timwekkenbc) [inline app/components/ScenariosManager/ScenarioGenerator/ScenarioGenerator.tsx]: `{ ...simulationContext, rulemap: true }`

File: app/components/ScenariosManager/ScenarioTester/ScenarioTester.tsx [hunk 6]
```diff
@@ -1,27 +1,40 @@
...
 import styles from "./ScenarioTester.module.css";
 import { runDecisionsForScenarios } from "@/app/utils/api";
+import { Scenario } from "@/app/types/scenario";
 import useResponsiveSize from "@/app/hooks/ScreenSizeHandler";
+import { dollarFormat } from "@/app/utils/utils";
 interface ScenarioTesterProps {
+  scenarios: Scenario[];
   jsonFile: string;
   ruleContent?: DecisionGraphType;
   uploader?: boolean;
```
Comment (timwekkenbc) [inline app/components/ScenariosManager/ScenarioTester/ScenarioTester.tsx]: Unrelated to this PR, but no longer used?

File: app/components/InputStyler/ArrayFormatter.tsx [hunk 7]
```diff
@@ -146,52 +162,139 @@ export default function ArrayFormatter(
...
+  // Utility function to remove an item from an array
+  const removeItemFromArray = (arrayPath: string) => {
+    const currentArray = rawData ? getValueFromPath(arrayPath, rawData) || [] : [];
+    currentArray.pop();
+    const newData = { ...rawData, [arrayPath]: currentArray };
+    setRawData(newData);
+  };
+
   if (editable) {
     if (Array.isArray(parsedValue)) {
```
Comment (timwekkenbc) [inline app/components/InputStyler/ArrayFormatter.tsx]: This is an outdated critique since this is not part of this PR, but it tends to cleaner to use guard clauses for this sort of thing. Like https://medium.com/nirman-tech-blog/using-guard-clauses-for-cleaner-react-code-77360e648659

**Review by timwekkenbc — APPROVED** [review]
Looks great! The sorting and filtering is especially awesome!
I just have some minor comments.
And a couple other considerations:
- Rename components to be in line with tab names
- Sorting doesn't really work on mobile, so maybe should just be disabled there?

File: app/components/ScenariosManager/ScenariosManager.tsx (hunk 2 above)
Comment (brysonjbest) [reply app/components/ScenariosManager/ScenariosManager.tsx]: Minor name update to activetabkey

File: app/components/ScenariosManager/ScenarioTester/ScenarioTester.tsx (hunk 6 above)
Comment (brysonjbest) [reply app/components/ScenariosManager/ScenarioTester/ScenarioTester.tsx]: Unsure what you mean by this comment.

File: app/components/ScenariosManager/ScenarioGenerator/ScenarioGenerator.tsx (hunk 4 above)
Comment (brysonjbest) [reply app/components/ScenariosManager/ScenarioGenerator/ScenarioGenerator.tsx]: Removed as unnecessary since handled with tab change.

File: app/components/InputStyler/ArrayFormatter.tsx (hunk 7 above)
Comment (brysonjbest) [reply app/components/InputStyler/ArrayFormatter.tsx]: Cool! Refactored and will keep in mind going forward.

File: app/components/ScenariosManager/ScenarioTester/ScenarioTester.tsx (hunk 6 above)
Comment (timwekkenbc) [reply app/components/ScenariosManager/ScenarioTester/ScenarioTester.tsx]: I was specifically talking about the `uploader?: boolean;` property, but this didn't highlight that line specifically like I'd thought

Comment (brysonjbest) [reply app/components/ScenariosManager/ScenarioTester/ScenarioTester.tsx]: Awesome, thank you, apparently I'm blind haha. Removed as no longer in use in this file.
